### ISSUE TITLE
Make suggest API more consistent and improve documentation

### ISSUE THREAD

**Issue by himkt** [optuna/optuna#2939]
## Motivation

We want to make behaviors of suggesting API for integer parameters and floating-point parameters consistent.
Our focuses are summarized in the description.

### 1. Suggest APIs variant for floating-point

There are four methods available for the current suggest API for a floating-point parameter: `suggest_float`, `suggest_uniform`, `suggest_loguniform`, and `suggest_discrete_uniform`. On the other hand, there is one method for each integer and categorical parameter: `suggest_int` and `suggest_categorical`. We plan to unify methods for floating point parameter into `suggest_float` and deprecate `suggest_uniform`, `suggest_loguniform`, and `suggest_discrete_uniform`. Then, we can change the implementation of `suggest_uniform`, `suggest_loguniform`, and `suggest_discrete_uniform` so that they are wrapped by `suggest_float`.

### 2. Documentation about behavior about the upper endpoint of floating-point distribution

While suggest API for an integer parameter would return the upper endpoint of the parameter space, suggest API for a float parameter would **not** return the upper endpoint according to documentation. Although it is true for some situations (e.g. `TPESampler`), samplers could return an upper endpoint due to numerical error of a floating-point. We decided to update our documentation that an upper endpoint would be included as a result of suggest API for float.

### 3. Allowing `suggest_float` to add a parameter during optimization (to make consistent with `suggest_int`)

In the current implementation, behavior is different between integer and float when adding a parameter (e.g. `step`) to suggest API during an optimization process. An error occurs for float while it works in the integer. To make the behavior consistent, we plan to change `suggest_float` so that it works when adding an additional parameter during optimization.

**Comment by xadrianzetx** [comment]
Regarding the first point - if I understand correctly, TODOs here are:
*  Make `suggest_uniform`, `suggest_loguniform,` and  `suggest_discrete_uniform` private.
*  Leave public APIs for those, but `@deprecate` them, and make them call `suggest_float`
*  Wrap private `suggest_uniform`, `suggest_loguniform,` and  `suggest_discrete_uniform` in `suggest_float`
*  Find and replace all occurrences of `suggest_uniform`, `suggest_loguniform,` and  `suggest_discrete_uniform` in the repo and replace with `suggest_float` (although this can be done separately, since `@deprecate` is used.

Please let me know if this sounds good, @himkt and I will open PR.

**Comment by himkt** [comment]
@xadrianzetx 
Thank you so much for paying your attention to this issue!

> Now i see that we could just remove them entirely and call _suggest like here

We want to deprecate methods instead of deleting them. `Leave public APIs for suggest_uniform, suggest_loguniform, and suggest_discrete_uniform , but @deprecate them, and make them call suggest_float` is the just thing what we will do.

> Find and replace all occurrences of suggest_uniform, suggest_loguniform, and suggest_discrete_uniform in the repo and replace with suggest_float (although this can be done separately, since @deprecate is used.)

Exactly.

You make sense about the first task. Let's work on the task together. If you send a PR, I'd happy to review it!

**Comment by nzw0301** [comment]
Should we replace `suggest_float` with `suggest_uniform` in the example code on https://optuna.org/? But I don't know the source document of this page...

**Comment by nzw0301** [comment]
@himkt  Right. My comment was complatey wrong. Thank you for kindly check code on eoptuna.org.

**Comment by xadrianzetx** [comment]
@himkt I can also work on points 2 and 3 if thats ok with you.

**Comment by himkt** [comment]
@xadrianzetx I really appreciate your contribution, I'm looking forward to review PR! Do you have questions for working on the tasks?

**Comment by xadrianzetx** [comment]
Thanks @himkt! Quick question - a code example for third point would be something like that, right?
```python
import optuna
from optuna.trial import Trial


def objective(trial: Trial):
    if trial.number % 2 == 0:
        x = trial.suggest_float("x", 1, 5)
    else:
        x = trial.suggest_float("y", 1, 5, step=2)
    return x ** 2


study = optuna.create_study()
study.optimize(objective, n_trials=10)
```

**Comment by himkt** [comment]
@xadrianzetx You're right. I think `suggest_int` would help to understand.
https://github.com/optuna/optuna/blob/16ddc8a390129c049c89ab26391cdc7bbb3351f9/optuna/trial/_trial.py#L331-L346

However, to complete (3), it needs a new unified distribution for floating point (like `IntUniformDistribution` for integer). `FloatDistribution` would be introduced by @nyanhi (https://github.com/optuna/optuna/pull/3063). So, for the first task, point (2) would be nice.

This issue is related to https://github.com/optuna/optuna/issues/2941. I think we can collaborate with @nyanhi and @keisuke-umezawa. Let me put comment on #2941 before working on (3).

So, this time, could you please start with (2)? :eyes:

**Comment by xadrianzetx** [comment]
Great explanation, thank you!
> So, this time, could you please start with (2)?

Sure!

**Comment by xadrianzetx** [comment]
#3063 has been merged, so we can move on to the last task from this list, right @himkt?

**Comment by himkt** [comment]
@xadrianzetx Right, let's go on to the final step! 🚀

**Comment by HideakiImamura** [comment]
[note] This issue will be closed in #3113.

**Comment by HideakiImamura** [comment]
Let me close this issue since we have finished all of tasks. Thanks @xadrianzetx, @nzw0301 and @himkt for long running work!
